### PR TITLE
🐛 Fixes is3793: error while pulling images 

### PR DESCRIPTION
**Pull request by pcrespov**
#### What cause the exception and why such a strange error message?
The error message reported in #3793: ``[...] 784 finished with exception: ''93dac263047b''``  is due to a ``KeyError``. And the trace shows the location:
```log
[... ]
 /simcore_service_dynamic_sidecar/core/docker_utils.py", line 129, in _parse_docker_pull_progress
    _, layer_total_size = all_image_pulling_data[image_name][layer_id]
[... ]
```
Since a ``KeyError `` converted to a string is directly the erroneous key 
```python
x = {}
try:
   y = x["93dac263047b"]
except KeyError as err:
  print(f"The error is {err}")
```
will print
```cmd
The error is 93dac263047b
```

#### How was it solved?

- 🐛Essentially initializing the item as ``all_image_pulling_data[image_name].setdefault(layer_id, (0,0))``. The former version of the algorithm expected a particular sequence of states that write and read from that container and **occasionally** that sequence was broken. 
- ♻️  On top this PR does a bit of refactoring of ``_pull_image_with_progress`` simplifying the code and documenting/annotating all parts more.

#### Extra 
- ♻️ improves error handling of deserialized session tokens (we experience some old tokens caused problems during login)

## Related issue/s

- Fixes ITISFoundation/osparc-simcore#3793

## How to test

- ``test_issue_3793_pulling_images_raises_error`` could **sometimes** reproduce  locally the error. It consumes a lot of time and disk so it has been excluded from CI and remains only for local manual test

## Checklist

- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [x] Documentation reflects the changes

### CODE REVIEW

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py [hunk 1]
```diff
@@ -94,58 +95,98 @@ async def pull_images(
...
+ImageName = str
+_ImagesInfoDict = dict[ImageName, _LayersInfoDict]
+
+
+class _ProgressDetailDict(TypedDict, total=False):
+    current: int
+    total: int
+
+
+class _DockerProgressDict(TypedDict, total=False):
```
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py:114]: ok just wondering if you would maybe look into packages/models-library/src/models_library/generated_models/docker_rest_api.py, you might find these models already

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py [hunk 2]
```diff
@@ -94,58 +95,98 @@ async def pull_images(
...
                     _DOWNLOAD_RATIO * layer_total_size
                     + (1 - _DOWNLOAD_RATIO)
                     * docker_pull_progress["progressDetail"]["current"]
                 ),
                 layer_total_size,
             )
-        elif docker_pull_progress["status"] == "Pull complete":
-            _, layer_total_size = all_image_pulling_data[image_name][layer_id]
-            all_image_pulling_data[image_name][layer_id] = (
+        elif status == _TargetPullStatus.PULL_COMPLETE:
```
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py:176]: ok not sure I get the whole issue. so sometimes the pull_complete or download_complete or extracting would come without downloading? is that the issue?

File: services/dynamic-sidecar/tests/unit/test_core_docker_utils.py [hunk 3]
```diff
@@ -129,3 +131,64 @@ def test_get_docker_service_images(compose_spec_yaml: str):
...
+                REGISTRY_USER="",
+                REGISTRY_PW="",
+                REGISTRY_SSL=False,
+            ),
+            progress_cb=_print_progress,
+            log_cb=_print_log,
+        )
+
+
+@pytest.mark.parametrize("repeat", ["first-pull", "repeat-pull"])
```
Comment (sanderegg) [inline services/dynamic-sidecar/tests/unit/test_core_docker_utils.py:173]: cool!

**Review by sanderegg — APPROVED** [review]
great and thanks a lot for the fix!

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py (hunk 1 above)
Comment (pcrespov) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py:114]: aha, i did not look carefully. I will check again.

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py (hunk 2 above)
Comment (pcrespov) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py:176]: the idea is that in some cases, the code READS first and in other WRITES to ``all_image_pulling_data``. But if it happens to read before it writes, then it raises a KeyError

**Review by GitHK — APPROVED** [review]
👍

**Pull request merged**